File: src/TMGToolbox/assignment/createTrafficDemand.py
"""
    Copyright 2022 Travel Modelling Group, Department of Civil Engineering, University of Toronto

    This file is part of TMGToolbox for Aimsun.

    TMGToolbox for Aimsun is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    TMGToolbox for Aimsun is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with TMGToolbox for Aimsun.  If not, see <http://www.gnu.org/licenses/>.
"""

from PyANGBasic import *
from PyANGKernel import *
from PyANGConsole import *
from PyANGDTA import *
import commonModule as CM

def create_traffic_demand(model, system, xtmf_parameters):
    """
    Function which generates and creates the TrafficDemand Data
    # to find a list of all folders qthelp://aimsun.com.aimsun.22.0/doc/UsersManual/ScriptExample1.
    """
    # the name of the folder object we will be creating
    folderName = "GKModel::trafficDemand"
    # add info from the OD Matrix into a traffic demand item which is used in the model
    trafficDemand = system.newObject("GKTrafficDemand", model)

    # build traffic demand folder is required
    CM.build_folder(model, trafficDemand, folderName)
    
    scheduleDemandItem = GKScheduleDemandItem()
    sectionType = model.getType("GKODMatrix") 
   
    for item in xtmf_parameters["demandParams"]:
        odMatrix = model.getCatalog().findObjectByExternalId(item["NameODMatrix"], sectionType)   
        # TODO make these parameters for the scenario length
        scheduleDemandItem.setFrom(int(item["InitialTime"] * 60.0))
        scheduleDemandItem.setDuration(int(item["Duration"] * 60.0))
        scheduleDemandItem.setTrafficDemandItem(odMatrix)
        trafficDemand.addToSchedule(scheduleDemandItem)
    
    trafficDemand.setName(xtmf_parameters["demandName"])
    
def run_xtmf(parameters, model, console):
    """
    A general function called in all python modules called by bridge. Responsible
    for extracting data and running appropriate functions.
    """
    # extract the parameters and save to dictionary
    xtmf_parameters = {
        "demandName": parameters["demandObjectName"],
        "demandParams": parameters["demandParams"],
    }
    _execute(model, console, xtmf_parameters)
    
def _execute(inputModel, console, xtmf_parameters):
    """
    Main execute function to run the simulation
    """
    model = inputModel
    system = GKSystem.getSystem()
    catalog = model.getCatalog()

    # delete the existing traffic demand objects
    CM.deleteAimsunObject(model, "GKTrafficDemand", xtmf_parameters["demandName"])

    # method to create the traffic demand object
    create_traffic_demand(model, system, xtmf_parameters)

File: src/TMG.Aimsun/assignment/CreateTrafficDemand.cs
/*
    Copyright 2022 Travel Modelling Group, Department of Civil Engineering, University of Toronto

    This file is part of TMGToolbox for Aimsun.

    TMGToolbox for Aimsun is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    TMGToolbox for Aimsun is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with TMGToolbox for Aimsun.  If not, see <http://www.gnu.org/licenses/>.
*/

using System;
using XTMF;
using TMG.Input;
using System.IO;

namespace TMG.Aimsun.assignment
{
    [ModuleInformation(Description = "Module to generate and build the Aimsun Traffic Demand object")]
    public class CreateTrafficDemand : IAimsunTool
    {
        public const string ToolName = "assignment/createTrafficDemand.py";

        [RunParameter("TrafficDemandName", "", "The new name of the traffic demand object created.")]
        public string demandName;

        [ModuleInformation(Description = "Module to Add the inputs for actually creating multiple traffic demand objects")]
        public class DemandParameters : IModule
        {
            [RunParameter("NameODMatrix", "", "Name of OD Matrix used to create the demand matrix from", Index = 0)]
            public string NameODMatrix;
            [RunParameter("InitialTime", 0.0, "Initial Starting Time in minutes", Index = 1)]
            public double InitialTime;
            [RunParameter("Duration", 0.0, "Duration of demand in minutes", Index = 2)]
            public double Duration;

            public string Name { get; set; }
            public float Progress => 0f;
            public Tuple<byte, byte, byte> ProgressColour => new Tuple<byte, byte, byte>(120, 25, 100);
            public bool RuntimeValidation(ref string error)
            {
                return true;
            }
        };

        [SubModelInformation(Description = "Traffic Demand Matrix Parameters", Required = true)]
        public DemandParameters[] DemandParams;

        public string Name { get; set; }

        public float Progress { get; set; }

        public Tuple<byte, byte, byte> ProgressColour => new Tuple<byte, byte, byte>(120, 25, 100);

        public bool RuntimeValidation(ref string error)
        {
            return true;
        }

        public bool Execute(ModellerController aimsunController)
        {
            if (aimsunController == null)
            {
                throw new XTMFRuntimeException(this, "AimsunController is not properly setup or initalized.");
            }
            return aimsunController.Run(this, ToolName,
                JsonParameterBuilder.BuildParameters(writer =>
                {
                    writer.WritePropertyName("demandObjectName");
                    writer.WriteValue(demandName);
                    writer.WritePropertyName("demandParams");
                    writer.WriteStartArray();
                    for (int i = 0; i < DemandParams.Length; i++)
                    {
                        writer.WriteStartObject();
                        writer.WritePropertyName("NameODMatrix");
                        writer.WriteValue(DemandParams[i].NameODMatrix);
                        writer.WritePropertyName("InitialTime");
                        writer.WriteValue(DemandParams[i].InitialTime);
                        writer.WritePropertyName("Duration");
                        writer.WriteValue(DemandParams[i].Duration);
                        writer.WriteEndObject();
                    }
                    writer.WriteEndArray();
                }));
        }
    }
}
